crates/uv-resolver/src/requires_python.rs

```diff
@@ -283,6 +283,87 @@ impl RequiresPython {
+    /// make. Generally, this means it should not be used inside the resolver,
+    /// but instead near the boundaries of the system (like formatting error
+    /// messages and writing the lock file). The reason for this is that
+    /// this simplification fundamentally changes the meaning of the marker,
+    /// and the *only* correct way to interpret it is in a context in which
+    /// `requires-python` is known to be true. For example, when markers from
+    /// a lock file are deserialized and turned into a `ResolutionGraph`, the
+    /// markers are "complexified" to put the `requires-python` assumption back
+    /// into the marker explicitly.
+    pub(crate) fn simplify_markers(&self, marker: MarkerTree) -> MarkerTree {
+        let simplified = marker.simplify_python_versions(Range::from(self.range().clone()));
+        // FIXME: This is a hack to avoid the hidden state created by
+        // ADD's `restrict_versions`. I believe this is sound, but it's
+        // wasteful and silly.
+        simplified
+            .try_to_string()
+            .map(|s| s.parse().unwrap())
+            .unwrap_or(MarkerTree::TRUE)
+    }
+
+    /// The inverse of `simplify_markers`.
+    ///
+    /// This should be applied near the boundaries of uv when markers are
+    /// deserialized from a context where `requires-python` is assumed. For
+    /// example, with `requires-python = '>=3.8'` and a marker like:
+    ///
+    /// ```text
+    /// python_full_version < '3.12'
+    /// ```
+    ///
+    /// It will be "complexified" to:
+    ///
+    /// ```text
+    /// python_full_version >= '3.8' and python_full_version < '3.12'
+    /// ```
+    pub(crate) fn complexify_markers(&self, mut marker: MarkerTree) -> MarkerTree {
+        // PERF: There's likely a way to amortize this, particularly
+        // the construction of `to_marker_tree`. But at time of
+        // writing, it wasn't clear if this was an actual perf problem
+        // or not. If it is, try a `std::sync::OnceLock`.
+        marker.and(self.to_marker_tree());
+        marker
+    }
+
+    /// Returns this `Requires-Python` specifier as an equivalent marker
+    /// expression utilizing the `python_version` marker field.
+    ///
+    /// This is useful for comparing a `Requires-Python` specifier with
+    /// arbitrary marker expressions. For example, one can ask whether the
+    /// returned marker expression is disjoint with another marker expression.
+    /// If it is, then one can conclude that the `Requires-Python` specifier
+    /// excludes the dependency with that other marker expression.
+    ///
+    /// If this `Requires-Python` specifier has no constraints, then this
+    /// returns a marker tree that evaluates to `true` for all possible marker
+    /// environments.
+    pub fn to_marker_tree(&self) -> MarkerTree {
+        let mut marker = self.range.lower().to_marker_tree(true);
+        marker.and(self.range.upper().to_marker_tree(false));
+        marker
+    }
```

Comment:
Note that this seems redundant with the recently added `markers` method. I tried using `markers` instead, but it actually resulted in a bunch of test failures. There are some subtle differences between my implementation (which revives it from older code) and the new `markers` method that I haven't investigated yet. It looks like it might come down to `python_full_version` versus `python_version`. cc @charliermarsh

Reply:
My implementation actually had a crucial bug in it (#6902).

Reply:
Can you try removing my impl's usages with this method? (Or use my impl?) Just a brief try after rebase would be appreciated. If you still see failures, let me know.

Reply:
Yup, nice fix. It works.